File: modules/python_workloads/mem/mem_retained_growth.py
"""One-shot retained-growth workload: the heap profile attributes the
whole live heap to the growth site."""

from prof_dump import run_profiled
from python_workloads.retained_growth import grow, retained_bytes
from python_workloads.workload_n import workload_n

DEFAULT_CHUNKS = 65536
CHUNK_BYTES = 1024


def main() -> None:
    chunks = workload_n(DEFAULT_CHUNKS)
    retained, out = run_profiled(lambda: grow(chunks, CHUNK_BYTES))
    print(f"retained {retained_bytes(retained)} bytes in {len(retained)} chunks; heap profile: {out}")


if __name__ == "__main__":
    main()

File: modules/java_workloads/mem/MemRetainedGrowth.java
// One-shot retained-growth workload: the allocation profile attributes
// the whole live heap to the growth site.

package monorepo.workloads;

import java.lang.ref.Reference;

public final class MemRetainedGrowth {
    private static final int DEFAULT_CHUNKS = 65_536;
    private static final int CHUNK_BYTES = 1024;

    private MemRetainedGrowth() {
    }

    public static void main(String[] args) throws Exception {
        int chunks = WorkloadN.workloadN(DEFAULT_CHUNKS);
        ProfDump.start();
        byte[][] retained = RetainedGrowth.grow(chunks, CHUNK_BYTES);
        String out = ProfDump.dump();
        System.out.printf("retained %d bytes in %d chunks; heap recording: %s%n",
            RetainedGrowth.retainedBytes(retained), retained.length, out);
        Reference.reachabilityFence(retained);
    }
}
